cocos2d/actions/animate.js

```diff
@@ -0,0 +1,363 @@
+
+let AnimateAction = cc.Class({
+    name: 'AnimateAction',
+    extends: cc.ActionInterval,
+
+    ctor (duration, props, opts) {
+        this._opts = opts = opts || Object.create(null);
+        this._props = Object.create(null);
+
+        // global easing or progress used for this action
+        opts.progress = opts.progress || this.progress;
+        if (opts.easing && typeof opts.easing === 'string') {
+            opts.easing = cc.easing[opts.easing];
+        }
+
+        for (let name in props) {
+            let value = props[name];
+
+            // property may have custom easing or progress function
+            let easing, progress;
+            if (value.value && (value.easing || value.progress)) {
+                easing = typeof value.easing === 'string' ? cc.easing[value.easing] : value.easing;
+                progress = value.progress;
+                value = value.value;
+            }
+
+            let isNumber = typeof value === 'number';
+            if (!isNumber && (!value.lerp || (!value.add && !value.mul) || !value.clone)) {
+                cc.warn(`Can not animate ${name} property, because it do not have [lerp, (add|mul), clone] function.`);
+                continue;
+            }
+
+            let prop = Object.create(null);
+            prop.value = value;
+            prop.easing = easing;
+            prop.progress = progress;
+            this._props[name] = prop;
+        }
+
+        this._originProps = props;
+		this.initWithDuration(duration);
+    },
+
+    clone () {
+        var action = new AnimateAction(this._duration, this._originProps, this._opts);
+        this._cloneDecoration(action);
+        return action;
+    },
+
+    startWithTarget (target) {
+        cc.ActionInterval.prototype.startWithTarget.call(this, target);
+
+        let relative = !!this._opts.relative;
+        let props = this._props;
+        for (let name in props) {
+            let value = target[name];
+            let prop = props[name];
+            
+            if (typeof value === 'number') {
+                prop.start = value;
+                prop.current = value;
+                prop.end = relative ? value + prop.value : prop.value;
+            }
+            else {
+                prop.start = value.clone();
+                prop.current = value.clone();
+                prop.end = relative ? (value.add || value.mul).call(value, prop.value) : prop.value;
```

Comment:
这个地方的处理感觉有点随意啊，对 props 的兼容性可能会有问题，另外，value.add 和 mul 也可能都不存在

Reply:
ctor 里有检测类型啦！

```js
            if (!isNumber && (!value.lerp || (!value.add && !value.mul) || !value.clone)) {
                 cc.warn(`Can not animate ${name} property, because it do not have [lerp, (add|mul), clone] function.`);
                 continue;
             }
```